src/platform/posix/posix_pollq_epoll.c

```diff
@@ -0,0 +1,372 @@
+		return (NNG_ESTATE);
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	if (pq->close) {
+		// This shouldn't happen!
+		nni_mtx_unlock(&pq->mtx);
+		return (NNG_ECLOSED);
+	}
+
+	node->pq     = pq;
+	node->events = 0;
+
+	// notifications disabled to begin with
+	ev.events   = 0;
+	ev.data.ptr = node;
+
+	rv = epoll_ctl(pq->epfd, EPOLL_CTL_ADD, node->fd, &ev);
+	if (rv != 0) {
+		rv = nni_plat_errno(errno);
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+	return (rv);
+}
+
+// common functionality for nni_posix_pollq_remove() and nni_posix_pollq_fini()
+// called while pq's lock is held
+static void
+nni_posix_pollq_remove_helper(nni_posix_pollq *pq, nni_posix_pollq_node *node)
+{
+	int                rv;
+	struct epoll_event ev;
+
+	node->events = 0;
+	node->pq     = NULL;
+
+	ev.events   = 0;
+	ev.data.ptr = node;
+
+	rv = epoll_ctl(pq->epfd, EPOLL_CTL_DEL, node->fd, &ev);
+	if (rv != 0) {
+		NNI_ASSERT(errno == EBADF || errno == ENOENT);
+	}
+}
+
+// nni_posix_pollq_remove removes the node from the pollq, but
+// does not ensure that the pollq node is safe to destroy.  In particular,
+// this function can be called from a callback (the callback may be active).
+void
+nni_posix_pollq_remove(nni_posix_pollq_node *node)
+{
+	nni_posix_pollq *pq = node->pq;
+
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	nni_posix_pollq_remove_helper(pq, node);
+
+	if (pq->close) {
+		nni_cv_wake(&pq->cv);
+	}
+	nni_mtx_unlock(&pq->mtx);
+}
+
+// nni_posix_pollq_init merely ensures that the node is ready for use.
+// It does not register the node with any pollq in particular.
+int
+nni_posix_pollq_init(nni_posix_pollq_node *node)
+{
+	NNI_ARG_UNUSED(node);
+	return (0);
+}
+
+// nni_posix_pollq_fini does everything that nni_posix_pollq_remove does,
+// but it also ensures that the callback is not active, so that the node
+// may be deallocated.  This function must not be called in a callback.
+void
+nni_posix_pollq_fini(nni_posix_pollq_node *node)
+{
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+	while (pq->active == node) {
+		pq->wait = node;
+		nni_cv_wait(&pq->cv);
+	}
+
+	nni_posix_pollq_remove_helper(pq, node);
+
+	if (pq->close) {
+		nni_cv_wake(&pq->cv);
+	}
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_arm(nni_posix_pollq_node *node, int events)
+{
+	int                rv;
+	struct epoll_event ev;
+	nni_posix_pollq *  pq = node->pq;
+
+	NNI_ASSERT(pq != NULL);
+	if (events == 0) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	node->events |= events;
+	ev.events   = node->events | NNI_EPOLL_FLAGS;
+	ev.data.ptr = node;
+
+	rv = epoll_ctl(pq->epfd, EPOLL_CTL_MOD, node->fd, &ev);
+	NNI_ASSERT(rv == 0);
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+void
+nni_posix_pollq_disarm(nni_posix_pollq_node *node, int events)
+{
+	struct epoll_event ev;
+
+	nni_posix_pollq *pq = node->pq;
+	if (pq == NULL) {
+		return;
+	}
+
+	nni_mtx_lock(&pq->mtx);
+
+	node->events &= ~events;
+	if (node->events == 0) {
+		ev.events = 0;
```

Comment:
I wonder if we actually need to make the system call in this case.  We have the ONESHOT flag, so we shouldn't keep getting called here....

Reply:
the only thing i can think of is the case in which an event has been armed, and disarm is called before any events have been delivered. this could result in the next event being delivered despite disarm having been called, since we wouldn't have actively disabled it.

if callers can deal with that, then agreed we should be able to avoid the call to `epoll_ctl()` here.

existing tests pass with your proposed change - let me know which you prefer.